rustler/src/resource.rs

```diff
@@ -145,9 +144,41 @@ where
         }
     }
 
+    /// Make a resource binary associated with the given resource
+    ///
+    /// The closure `f` is called with the referenced object and must return a slice with the same
+    /// lifetime as the object. This means that the slice either has to be derived directly from
+    /// the instance or that it has to have static lifetime.
+    pub fn make_binary<'env, 'a, F>(&self, env: Env<'env>, f: F) -> Binary<'env>
+    where
+        F: FnOnce(&'a T) -> &'a [u8],
+    {
+        unsafe { self.make_binary_unsafe(env, f) }
+    }
+
+    /// Make a resource binary without strict lifetime checking
+    ///
+    /// The user *must* ensure that the lifetime of the returned slice is at least as long as the
+    /// lifetime of the referenced instance.
+    pub unsafe fn make_binary_unsafe<'env, 'a, 'b, F>(&self, env: Env<'env>, f: F) -> Binary<'env>
+    where
+        F: FnOnce(&'a T) -> &'b [u8],
```

Comment:
What if the unsafe version does not have the function and directly receives the slice? This way I can invoke it directly for my case where I have to generate multiple resource binaries from a loop.

Reply:
> > I think adding a `fn make_binary_unsafe<'env, 'a>(&self, env: Env<'env>, data: &'a [u8]) -> Binary` might be better, if we don't enforce the lifetime, I don't think there would be any need to have it go through a callback?
> 
> Having the same signature helps to clarify the difference, IMO. It also makes it clear that the binary's lifetime _should_ be connected to the referenced object.

I mentioned this earlier, and I am a bit undecided on it, but @filmor has a good point. Having it have mostly the same signature might help show that these two are equivalent. What do you think?

Note that you can still call the unsafe variant and return a slice from the outer scope since the lifetimes are not tied in the unsafe variant. It shouldn't produce any worse machine code.

Reply:
I see, I thought I would I would not be able to call the outer scope. If I can, then perfect! :+1: